File: 2015/13/python/main.py
from pathlib import Path
from typing import Dict, List
from itertools import permutations

def parse(data:List[str]) -> Dict[str, Dict[str,int]]:
    out_dict = {}
    for r in data:
        vals = r.split()
        p1 = vals[0]
        p2 = vals[-1].strip(".")
        hap = int(vals[3])

        if vals[2] == "lose":
            hap = -hap

        if p1 not in out_dict:
            out_dict[p1] = {}
            
        out_dict[p1].update({p2:hap})
    return out_dict

def parse_2(data:List[str]) -> Dict[str, Dict[str,int]]:
    out_dict = {}
    for r in data:
        vals = r.split()
        p1 = vals[0]
        p2 = vals[-1].strip(".")
        hap = int(vals[3])

        if vals[2] == "lose":
            hap = -hap

        if p1 not in out_dict:
            out_dict[p1] = {}

        out_dict[p1].update({p2:hap})
        all_guests = set(out_dict.keys()).union(
        {guest for subdict in out_dict.values() for guest in subdict.keys()})
    
        for guest in all_guests:
            if guest not in out_dict:
                out_dict[guest] = {}
            out_dict[guest]["you"] = 0
        
    out_dict["you"] = {guest: 0 for guest in all_guests}
    return out_dict

def validate_seating(arrangement, preferences):
    total_happiness = 0
    n = len(arrangement)

    for i in range(n):
        current_person = arrangement[i]
        left_neighbor = arrangement[(i - 1) % n]
        right_neighbor = arrangement[(i + 1) % n]

        # Check if the neighbors exist in the preferences
        if left_neighbor in preferences[current_person]:
            total_happiness += int(preferences[current_person][left_neighbor])
        else:
            return None

        if right_neighbor in preferences[current_person]:
            total_happiness += int(preferences[current_person][right_neighbor])
        else:
            return None

    return total_happiness

def find_best_seating(preferences):
    people = list(preferences.keys())
    max_happiness = None

    for arrangement in permutations(people):
        happiness = validate_seating(arrangement, preferences)
        if happiness is not None:
            if max_happiness is None or happiness > max_happiness:
                max_happiness = happiness

    return max_happiness




def main() -> None:
    data = Path("../input.txt").read_text().strip().splitlines()
    example = Path("../example.txt").read_text().strip().splitlines()
    values = parse(data)
    print("=" * 20, " Part 1 ", "=" * 20)
    print(f"Max happiness score: {find_best_seating(values)}")
    print("=" * 20, " Part 2 ", "=" * 20)
    values = parse_2(data)
    print(f"Max happiness score: {find_best_seating(values)}")

    
if __name__ == "__main__":
    main()

File: 2015/13/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Preferences map[string]map[string]int

func parse(data []string) Preferences {
	outDict := make(Preferences)

	for _, line := range data {
		parts := strings.Fields(line)
		person1 := parts[0]
		person2 := strings.TrimSuffix(parts[len(parts)-1], ".")
		happiness, _ := strconv.Atoi(parts[3])
		if parts[2] == "lose" {
			happiness = -happiness
		}

		if _, exists := outDict[person1]; !exists {
			outDict[person1] = make(map[string]int)
		}
		outDict[person1][person2] = happiness
	}
	return outDict
}

func parse2(data []string) Preferences {
	preferences := parse(data)
	allGuests := make(map[string]bool)

	for person, prefs := range preferences {
		allGuests[person] = true
		for guest := range prefs {
			allGuests[guest] = true
		}
	}

	preferences["you"] = make(map[string]int)
	for guest := range allGuests {
		preferences["you"][guest] = 0
		if _, exists := preferences[guest]; !exists {
			preferences[guest] = make(map[string]int)
		}
		preferences[guest]["you"] = 0
	}
	return preferences
}

func calculateHappiness(arrangement []string, preferences Preferences) int {
	totalHappiness := 0
	numPeople := len(arrangement)

	for i := 0; i < numPeople; i++ {
		person := arrangement[i]
		leftNeighbor := arrangement[(i-1+numPeople)%numPeople]
		rightNeighbor := arrangement[(i+1)%numPeople]

		totalHappiness += preferences[person][leftNeighbor] + preferences[person][rightNeighbor]
	}
	return totalHappiness
}

func findBestSeating(preferences Preferences) int {
	people := []string{}
	for person := range preferences {
		people = append(people, person)
	}

	permutations := generatePermutations(people)
	maxHappiness := -1 << 31 // Minimum int value

	for _, arrangement := range permutations {
		if happiness := calculateHappiness(arrangement, preferences); happiness > maxHappiness {
			maxHappiness = happiness
		}
	}
	return maxHappiness
}

func generatePermutations(arr []string) [][]string {
	var res [][]string
	var backtrack func(int)

	backtrack = func(start int) {
		if start == len(arr) {
			dst := make([]string, len(arr))
			copy(dst, arr)
			res = append(res, dst)
			return
		}
		for i := start; i < len(arr); i++ {
			arr[start], arr[i] = arr[i], arr[start]
			backtrack(start + 1)
			arr[start], arr[i] = arr[i], arr[start]
		}
	}

	backtrack(0)
	return res
}

func main() {
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Println("Failed to read file:", err)
		return
	}
	defer file.Close()

	var data []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		data = append(data, scanner.Text())
	}

	sep := strings.Repeat("=", 20)
	part1 := findBestSeating(parse(data))
	part2 := findBestSeating(parse2(data))

	fmt.Printf("%s Part 1 %s \nMax happiness score: %d\n", sep, sep, part1)
	fmt.Printf("%s Part 2 %s \nMax happiness score: %d\n", sep, sep, part2)
}
